File: dotnet/tests/IForevents.Tests/MockApi.cs
using System.Net;
using System.Text;
using System.Text.Json;

namespace IForevents.Tests;

/// <summary>Records every request and answers like the real api; a scenario injects faults. Plugged in as an HttpMessageHandler, so no sockets.</summary>
public sealed class MockApi : HttpMessageHandler
{
    public sealed record Recorded(string Path, Dictionary<string, string> Headers, JsonElement Body)
    {
        public string? Header(string name) => Headers.TryGetValue(name.ToLowerInvariant(), out var v) ? v : null;
        public List<string> EventNames() => Body.GetProperty("events").EnumerateArray().Select(e => e.GetProperty("name").GetString()!).ToList();
    }

    public readonly List<Recorded> Requests = new();
    public Func<Recorded, (int status, object body, Dictionary<string, string>? headers)?>? Scenario;
    public const string BaseUrl = "http://mock.iforevents.test";

    public HttpClient Client() => new HttpClient(this, disposeHandler: false);

    public List<Recorded> ByPath(string path)
    {
        lock (Requests) return Requests.Where(r => r.Path == path).ToList();
    }

    protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
    {
        var text = request.Content == null ? "{}" : await request.Content.ReadAsStringAsync(cancellationToken);
        var headers = new Dictionary<string, string>();
        foreach (var h in request.Headers) headers[h.Key.ToLowerInvariant()] = string.Join(",", h.Value);
        if (request.Content != null) foreach (var h in request.Content.Headers) headers[h.Key.ToLowerInvariant()] = string.Join(",", h.Value);
        var rec = new Recorded(request.RequestUri!.AbsolutePath, headers, JsonDocument.Parse(text).RootElement.Clone());
        lock (Requests) Requests.Add(rec);
        var answer = Scenario?.Invoke(rec);
        if (answer == null)
        {
            if (rec.Header("X-Project-Key") != "pk_test") answer = (401, new { error = "invalid project key" }, null);
            else answer = rec.Path switch
            {
                "/v1/events/identify" => (201, new { user = new { uuid = "11111111-1111-4111-8111-111111111111" } }, null),
                "/v1/events/track" => (201, new { status = "ok", user_uuid = "22222222-2222-4222-8222-222222222222" }, null),
                "/v1/events/batch" => (202, new { status = "queued", user_uuid = "33333333-3333-4333-8333-333333333333" }, null),
                _ => (404, new { error = "not found" }, null),
            };
        }
        var (status, body, extra) = answer.Value;
        var res = new HttpResponseMessage((HttpStatusCode)status) { Content = new StringContent(JsonSerializer.Serialize(body), Encoding.UTF8, "application/json") };
        if (extra != null) foreach (var h in extra) res.Headers.TryAddWithoutValidation(h.Key, h.Value);
        return res;
    }
}

File: php/iforevents/tests/MockApi.php
<?php

declare(strict_types=1);

namespace IForevents\Tests;

/**
 * Records every request and answers like the real api; a scenario closure
 * injects faults. Plugged in as APIConfig::$transport, so no sockets.
 */
final class MockApi
{
    /** @var list<array{path:string, headers:array<string,string>, body:array<string,mixed>}> */
    public array $requests = [];
    /** @var null|callable(array): ?array{0:int,1:array,2:array} */
    public $scenario = null;

    public function reset(): void
    {
        $this->requests = [];
        $this->scenario = null;
    }

    /** @return list<array{path:string, headers:array<string,string>, body:array<string,mixed>}> */
    public function byPath(string $path): array
    {
        return array_values(array_filter($this->requests, static fn (array $r) => $r['path'] === $path));
    }

    /** @return array{0:int,1:string,2:array<string,string>} */
    public function __invoke(string $url, string $payload, array $headers, float $timeout): array
    {
        $path = parse_url($url, PHP_URL_PATH) ?: '';
        $lower = [];
        foreach ($headers as $k => $v) {
            $lower[strtolower($k)] = $v;
        }
        $req = ['path' => $path, 'headers' => $lower, 'body' => json_decode($payload, true) ?? []];
        $this->requests[] = $req;
        $answer = $this->scenario !== null ? ($this->scenario)($req) : null;
        if ($answer === null) {
            if (($lower['x-project-key'] ?? '') !== 'pk_test') {
                $answer = [401, ['error' => 'invalid project key'], []];
            } else {
                $answer = match ($path) {
                    '/v1/events/identify' => [201, ['user' => ['uuid' => '11111111-1111-4111-8111-111111111111']], []],
                    '/v1/events/track' => [201, ['status' => 'ok', 'user_uuid' => '22222222-2222-4222-8222-222222222222'], []],
                    '/v1/events/batch' => [202, ['status' => 'queued', 'user_uuid' => '33333333-3333-4333-8333-333333333333'], []],
                    default => [404, ['error' => 'not found'], []],
                };
            }
        }
        [$status, $body, $resHeaders] = $answer;
        $norm = [];
        foreach ($resHeaders as $k => $v) {
            $norm[strtolower($k)] = (string) $v;
        }
        return [$status, json_encode($body), $norm];
    }
}
